### PR TITLE
question finder tile

### PR DESCRIPTION
**Pull request by hannah-whelan**
https://github.com/isaaccomputerscience/isaac-cs-issues/issues/181

### CODE REVIEW

File: src/app/services/api.ts [hunk 1]
```diff
@@ -303,6 +303,9 @@ export const api = {
     testFreeTextQuestion: (userDefinedChoices: Choice[], testCases: TestCaseDTO[]) => {
       return endpoint.post("/questions/test?type=isaacFreeTextQuestion", { userDefinedChoices, testCases });
     },
+    randomQuestions: (): AxiosPromise<ApiTypes.IsaacQuestionPageDTO[]> => {
+      return endpoint.get("/questions/randomQuestions");
```
Comment (jmgq) [inline src/app/services/api.ts:307]: This is more of an API comment, really, but it should be `random_questions`, since it seems like we tend to use underscores for URLs, rather than camel case.

File: src/app/components/elements/Dashboard.tsx [hunk 2]
```diff
@@ -69,15 +77,28 @@ export const Dashboard = ({
...
+        </Col>
+      )}
+      {isTeacherOrAbove(user) ? (
+        <PromoOrFeaturedNews contentType={promoItem ? "promo" : "news"} />
+      ) : (
+        <ShowLoading
+          until={questionData}
+          thenRender={(questionData) => (
+            <Col md="12" lg={expanded ? "12" : "7"}>
+              {questionData && <QuestionCard setExpanded={setExpanded} questionData={questionData} />}
```
Comment (jmgq) [inline src/app/components/elements/Dashboard.tsx:96]: Is `questionData &&` actually needed? I would have thought that if you reach this point it means that `questionData` won't be falsy.

File: src/mocks/data.ts [hunk 3]
```diff
@@ -3,6 +3,689 @@ import { DAYS_AGO } from "../test/utils";
 import { RegisteredUserDTO, IsaacEventPageDTO, UserSummaryWithGroupMembershipDTO } from "../IsaacApiTypes";
 import { EventBookingDTO } from "../IsaacApiTypes";
 
+export const mockRandomQuestions = [
```
Comment (jmgq) [inline src/mocks/data.ts:6]: The type is missing.

File: src/test/pages/IsaacApp.test.tsx [hunk 4]
```diff
@@ -162,7 +142,7 @@ describe("IsaacApp", () => {
     });
   });
 
-  let roles = [
+  const roles = [
```
Comment (jmgq) [inline src/test/pages/IsaacApp.test.tsx:145]: Same comment about the type here.

File: src/app/components/elements/cards/QuestionCard.tsx [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+interface QuestionCardProps {
+  setExpanded: (expanded: boolean) => void;
+  questionData: IsaacQuestionPageDTO[];
+}
+
+const QuestionCard = ({ setExpanded, questionData }: QuestionCardProps) => {
+  const [expandRef, updateExpandRef] = useStatefulElementRef<HTMLDivElement>();
+  const { expandButton, outerClasses, expanded } = useExpandContent(true, expandRef);
+  const [question, setQuestion] = useState<number>(0);
+  const isMoreQuestions = question < 5;
```
Comment (jmgq) [inline src/app/components/elements/cards/QuestionCard.tsx:28]: This could just be me, but `isMoreQuestions` sounds a bit strange. What about something like `moreQuestionsAvailable`? Or `moreQuestionsPending`?

File: src/test/components/elements/Dashboard.test.tsx [hunk 6]
```diff
@@ -32,18 +32,23 @@ describe("Dashboard", () => {
...
-  it("if TEACHER user is logged in and promo item is available, the promo tile is visible, and featured news tile is not", async () => {
-    setupTest("TEACHER");
-    const promoTile = await screen.findByTestId("promo-tile");
-    const featuredNewsTile = screen.queryByTestId("featured-news-item");
-    expect(promoTile).toBeInTheDocument();
-    expect(featuredNewsTile).toBeNull();
-    const promoTitle = screen.getByText(mockPromoItem.title);
-    expect(promoTitle).toBeInTheDocument();
-  });
+  let roles: TestUserRole[] = ["TEACHER", "EVENT_LEADER", "CONTENT_EDITOR", "EVENT_MANAGER", "ADMIN"];
```
[the remark this answers is not in the document]
Comment (hannah-whelan) [reply src/test/components/elements/Dashboard.test.tsx:35]: Oops yes this was due to the fact that "ANONYMOUS" is not a UserRole so rather than having `UserRole | "ANONYMOUS"` in some tests, I used TestUserRole. However in this case it's definitely not required.

File: src/app/services/api.ts (hunk 1 above)
Comment (hannah-whelan) [reply src/app/services/api.ts:307]: Agreed, happy to update this when the endpoint is renamed in API.

File: src/app/components/elements/cards/QuestionCard.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+);
+interface QuestionCardProps {
+  setExpanded: (expanded: boolean) => void;
+  questionData: IsaacQuestionPageDTO[];
+}
+
+const QuestionCard = ({ setExpanded, questionData }: QuestionCardProps) => {
+  const [expandRef, updateExpandRef] = useStatefulElementRef<HTMLDivElement>();
+  const { expandButton, outerClasses, expanded } = useExpandContent(true, expandRef);
+  const [question, setQuestion] = useState<number>(0);
```
Comment (jmgq) [inline src/app/components/elements/cards/QuestionCard.tsx:27]: After reading this file, it looks like this should be called `questionIndex`, or similar.

File: src/app/components/elements/cards/QuestionCard.tsx [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        {isMoreQuestions && (
+          <Button className="next-question bg-transparent border-0 btn-link" onClick={() => setQuestion(question + 1)}>
+            Next question
+          </Button>
+        )}
+      </Row>
+      <Card
+        className={classnames(outerClasses, expanded ? "random-question-panel" : "mt-2 pb-2")}
+        data-testid="question-content-card"
+        style={expanded ? { maxHeight: "560px" } : { maxHeight: "450px" }}
```
Comment (jmgq) [inline src/app/components/elements/cards/QuestionCard.tsx:47]: Small one, but given that the only thing that changes depending on the `expanded` variable is the maximum height, it would be a bit more readable to write it like this, I think:
```typescript
style={{ maxHeight: expanded ? "560px" : "450px" }}
```

What do you think?

The same goes for the style a few lines below this one.

File: src/app/components/elements/cards/QuestionCard.tsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        <CardBody className="p-3">
+          <div
+            id="question-body"
+            ref={updateExpandRef}
+            style={expanded ? { maxHeight: "500px" } : { maxHeight: "400px" }}
+            className="overflow-auto hidden-scrollbar"
+          >
+            {isMoreQuestions ? <IsaacContent doc={questionData[question]} /> : <GoToQuestionFinder />}
+          </div>
+          {expanded ? expandButton : isMoreQuestions && expandButton}
```
Comment (jmgq) [inline src/app/components/elements/cards/QuestionCard.tsx:58]: Maybe this is just me, but I find this a bit hard to read. Isn't this equivalent to:
```typescript
{expanded || isMoreQuestions ? expandButton : null}
```

File: src/test/components/elements/cards/QuestionCard.test.tsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    const checkAnswerButton = screen.getByRole("button", { name: "Check my answer" });
+    const expandButton = screen.getByRole("button", {
+      name: /expand content/i,
+    });
+    [nextQuestion, questionTile, question, checkAnswerButton, expandButton].forEach((element) => {
+      expect(element).toBeInTheDocument();
+    });
+  });
+
+  it("renders the question tile as minimised by default, and expands to panel when the expand button is clicked", async () => {
```
Comment (jmgq) [inline src/test/components/elements/cards/QuestionCard.test.tsx:40]: Rather than using `toHaveClass` to test this, can't you use `toBeVisible` instead? It would make it a lot more readable (and less likely to change in the future if the classes are renamed, although this is not as important).

File: src/app/components/elements/cards/QuestionCard.tsx (hunk 9 above)
Comment (hannah-whelan) [reply src/app/components/elements/cards/QuestionCard.tsx:58]: Perhaps this instead?

`{(expanded || isMoreQuestions) && expandButton}`

File: src/app/components/elements/cards/QuestionCard.tsx (hunk 8 above)
Comment (hannah-whelan) [reply src/app/components/elements/cards/QuestionCard.tsx:47]: Definitely, thanks

File: src/test/components/elements/cards/QuestionCard.test.tsx [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    const questionContents = screen.getByTestId("question-content-card");
+    expect(questionContents).not.toHaveClass("random-question-panel");
+    await clickButton("Expand content");
+    expect(questionTile).not.toHaveClass("question-tile");
+    expect(questionContents).toHaveClass("random-question-panel");
+  });
+
+  it("renders the next question if next question button is pressed", async () => {
+    await setupTest();
+    await clickButton("Next question");
```
Comment (jmgq) [inline src/test/components/elements/cards/QuestionCard.test.tsx:53]: In tests like these, I normally also tend to assert that the question is not in the document before I press the button, just to make the intention even clearer. It's just a suggestion, so feel free to leave it as it is.

File: src/app/components/elements/ScrollShadows.tsx [hunk 12]
```diff
@@ -47,10 +47,12 @@ export const ScrollShadows = <T extends HTMLElement>({ element }: { element: T |
   useEffect(() => {
     if (!element) return;
     const resizeObserver = new ResizeObserver((entries) => {
-      const element = entries[0].target;
-      setClientWidth(element.clientWidth);
-      setScrollWidth(element.scrollWidth);
-      setScrollLeft(element.scrollLeft);
+      setTimeout(() => {
```
Comment (jmgq) [inline src/app/components/elements/ScrollShadows.tsx:50]: Could you elaborate on this change, please? It looks a bit hacky, given that the timeout is 0.

File: src/app/components/elements/Dashboard.tsx (hunk 2 above)
Comment (jmgq) [reply src/app/components/elements/Dashboard.tsx:96]: In that case, the following should do the trick:
```typescript
thenRender={(questionData: IsaacQuestionPageDTO[]) => (
```

File: src/app/components/elements/cards/QuestionCard.tsx (hunk 9 above)
Comment (jmgq) [reply src/app/components/elements/cards/QuestionCard.tsx:58]: I tend to never use `condition && theThingThatWeWantToDisplay` because it relies on the fact that in JS, the `theThingThatWeWantToDisplay` will be returned if `condition` is truthy, but this is not how it works in most languages. Normally, in other languages, the result of this expression is a boolean. So I find it less readable. But if the way you have suggested is common practice in JS and any JS developer would understand it without problems, then go ahead. I just tend to error on the side of caution and write code as readable as possible.

File: src/app/components/elements/ScrollShadows.tsx (hunk 12 above)
Comment (hannah-whelan) [reply src/app/components/elements/ScrollShadows.tsx:50]: Yeah this one wasn't ideal. Essentially without this I get occasional "ResizeObserver loop limit exceeded" errors, when using the expand/close button. I did some googling and found it's a relatively common problem, here was one of the explanations: 

`The root of this problem seems to be React re-rendering (or any computationally intensive work, really) happening as a direct result of the resize event. If you defer that work to after the resize event has been handled, the browser is totally fine with it because it's not being prevented from delivering all of its event notifications in time.`

The most common workaround seems to be the setTimeout 0 option, so that React handles it at the end and avoids errors. Using requestAnimationFrame() was also a suggested fix but that one caused other side effects. Debounce/throttle were also suggested fixes. I've manually tested it quite a bit and this timeout seems to not have caused any unexpected behaviour, and it fixed the error.

I did see some suggestions that the error won't happen once we are using react-dom 18, but we are currently on 17.

Comment (jmgq) [reply src/app/components/elements/ScrollShadows.tsx:50]: Thank you for the detailed explanation! That makes sense. I think this is one of those rare occasions in which I would definitely recommend adding a comment with your explanation, because otherwise a developer won't be able to figure out why the timeout is needed just by looking at the code. If we already have a ticket for upgrading React, it would be good to mention in that ticket that we should revisit this then and check if we can remove the workaround.

File: src/app/components/elements/cards/QuestionCard.tsx (hunk 9 above)
Comment (hannah-whelan) [reply src/app/components/elements/cards/QuestionCard.tsx:58]: Yeah it's perfectly understandable in JS. I guess I have the (dubious!) advantage of having worked almost exclusively in JS so this feels like standard practise for me - I'll make sure to be aware of it in other languages though!

File: src/test/components/elements/cards/QuestionCard.test.tsx (hunk 10 above)
Comment (hannah-whelan) [reply src/test/components/elements/cards/QuestionCard.test.tsx:40]: `toBeVisible` wouldn't work here; pressing the expand button doesn't change the visibility or switch components, it just resizes some elements - as far as I know the only way to verify that is going to be checking the classes.

File: src/test/components/elements/cards/QuestionCard.test.tsx (hunk 11 above)
Comment (hannah-whelan) [reply src/test/components/elements/cards/QuestionCard.test.tsx:53]: Updated this one to assert that question 1 and not question 2 appears, then after button press question 1 is gone and question 2 appears.

File: src/mocks/data.ts (hunk 3 above)
Comment (hannah-whelan) [reply src/mocks/data.ts:6]: This is another area where I believe there are some complex typing problems in the codebase. `IsaacQuestionPageDTO[]` is the expected type here, but that type expects `children` to be `ContentDTO[]`. In reality in this instance children can be `QuestionDTO` (which includes `hints`). 

If I update it so that `children` is `(QuestionDTO | ContentDTO)[];`, then `children` in `hints` start objecting because that's typed to be `ContentBaseDTO[]`. If I change that to `ContentDTO[]` it then complains about `src` in some of the children of hints. `src` is only expected from `MediaDTO` but that would then mean making `children` to be `(QuestionDTO | ContentDTO | MediaDTO)[]` 

But theeeen it objects about `choices` which only comes from `IsaacMultiChoiceQuestionDTO`. It feels a bit like playing whack-a-mole.

Setting `children` of `ContentDTO` to this: `(QuestionDTO | ContentDTO | MediaDTO | IsaacMultiChoiceQuestionDTO)[];` appears to solve it, but that just feels wrong, particularly because i've only used multiple choice question as the test data. If I change that test data to what is returned from a different question type, I suspect it would throw up new errors. 

In summary... it works without setting the type, and unless I'm missing something obvious I think it's going to be a headache for a future ticket to fix these types.

File: src/app/services/api.ts (hunk 1 above)
Comment (hannah-whelan) [reply src/app/services/api.ts]: renamed endpoint to `/questions/random` as per API changes

File: src/mocks/data.ts (hunk 3 above)
Comment (jmgq) [reply src/mocks/data.ts:6]: Thanks for the detailed explanation! I think you're right, this sounds like something that should be solved in its own ticket.

Having said that, it sounds like the only cause of errors here is the `children` property. My question is: are the children of `IsaacQuestionPageDTO` **ALWAYS** `QuestionDTO`s? If so, then an easy solution is to add the following to `IsaacQuestionPageDTO`:

```typescript
children?: QuestionDTO[];
```
That way, you'll override the type of `children` for this `IsaacQuestionPageDTO` while leaving `ContentDTO` intact. Also, if `IsaacQuestionPageDTO` is always guaranteed to return a `children` property (even if it's an empty array), then you should remove the question mark.

But if this does not work, then please create a new ticket to handle this and ignore my comments.

File: src/app/components/elements/cards/QuestionCard.tsx (hunk 8 above)
Comment (jmgq) [reply src/app/components/elements/cards/QuestionCard.tsx]: There is another instance of this a few lines below this one.

File: src/test/components/elements/cards/QuestionCard.test.tsx (hunk 10 above)
Comment (jmgq) [reply src/test/components/elements/cards/QuestionCard.test.tsx:40]: No worries, it was worth a try :-)

File: src/mocks/data.ts (hunk 3 above)
Comment (hannah-whelan) [reply src/mocks/data.ts:6]: Unfortunately no, the children of `IsaacQuestionPageDTO` can be a variety of things. Just taking the one sample question that I used to build the mock data, it has a `ContentDTO` and `QuestionDTO` in the array. I'm assuming that may vary between different questions. If set it as this 

`children?: (ContentDTO | QuestionDTO)[];` 

Then it still objects about the children of `hints` because apparently `ContentBaseDTO` shouldn't have children, and those children appear to be an assortment of types like `ContentDTO`, `MediaDTO` (and likely more if I looked at more data). 

I've created a ticket for it now as it feels like one that's going to take some time to resolve: https://github.com/isaaccomputerscience/isaac-cs-issues/issues/237

File: src/app/components/elements/cards/QuestionCard.tsx (hunk 8 above)
Comment (jmgq) [reply src/app/components/elements/cards/QuestionCard.tsx]: I think you may have accidentally changed the value of the pixels, since they used to be 500/400.

Comment (hannah-whelan) [reply src/app/components/elements/cards/QuestionCard.tsx]: Sometimes I'm even impressed with my own idiocy 😁 fixed now, thanks

**Review by jmgq — APPROVED** [review]
